File: 874 - Walking Robot Simulation.py
class Solution:
    def robotSim(self, commands: "list[int]", obstacles: "list[list[int]]") -> int:
        p = [0, 0]
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        direc = 0
        obstacles = set(tuple(map(tuple, obstacles)))
        res = 0
        for com in commands:
            if com == -1:
                direc = (direc + 1) % 4
            elif com == -2:
                direc = (direc + 3) % 4
            else:
                for i in range(com):
                    t = (p[0] + directions[direc][0],
                         p[1] + directions[direc][1])
                    if t not in obstacles:
                        p = t
                        dist = p[0] ** 2 + p[1] ** 2
                        res = res if res > dist else dist
                    else:
                        break
        return res


s = Solution()
a = s.robotSim(commands=[4, -1, 4, -2, 4], obstacles=[[2, 4]])
print(a)

File: 874 - Walking Robot Simulation.cpp
#include <vector>
#include <set>
#include <unordered_set>


using namespace std;

class Solution {
public:
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
        int x = 0, y = 0, res = 0, d = 0, nx, ny, dist;
        int dir[][2] = { {0, 1}, {1, 0}, {0, -1}, {-1, 0} };
        set<pair<int, int>> s;
        for (vector<int>& o : obstacles) {
            s.insert(make_pair(o[0], o[1]));
        }
        for (int cmd : commands) {
            if (cmd == -2) {
                d = (d + 3) % 4;
            }
            else if (cmd == -1) {
                d = (d + 1) % 4;
            }
            else {
                for (int i = 0; i < cmd; i++) {
                    nx = x + dir[d][0];
                    ny = y + dir[d][1];
                    auto p = make_pair(nx, ny);
                    if (s.find(p) == s.end()) {
                        x = nx;
                        y = ny;
                        dist = x * x + y * y;
                        res = res > dist ? res : dist;
                    }
                    else {
                        break;
                    }
                }
            }
        }
        return res;
    }
};